### PR TITLE
Adding Hammersley point set along as tests and documentation

### LINKED ISSUE #572: Add Hammersley point set discrete distribution
## Motivation

QMCPy currently provides several low-discrepancy constructions, but
does not appear to provide a Hammersley point set distribution.

Hammersley point sets are a classical low-discrepancy construction
used in quasi-Monte Carlo numerical integration.

## Proposed feature

Add a Hammersley point set discrete distribution following the existing
QMCPy distribution interfaces and conventions.

The construction is based on

$x_n = (\frac{n}{N}, \phi_{b_1}(n), ..., \phi_{b_{d-1}}(n)),$

where $\phi_b$ denotes the radical inverse function in base b.

## Proposed implementation

- Add the new distribution to the appropriate QMCPy module.
- Follow the existing distribution API.
- Add unit tests for the mathematical construction and edge cases.
- Add documentation and examples.
- Add references to the relevant literature.

I would be interested in implementing this feature. I have previous
experience working with quasi-Monte Carlo numerical integration and
would be happy to work on it if the maintainers consider it a useful
addition.

### PR DESCRIPTION
**Pull request by Samuel-Vangu**
## Summary
- Issue: Add Hammersley point set discrete distribution #572
- Algorithmic or API impact: Adds a new `HammersleyPointSet` discrete distribution class. 
Purely additive: no existing class, method, or public API signature is modified. The implementation reuses `Halton` internally (constructing a `d-1` dimensional unrandomized `Halton` instance for the radical-inverse coordinates) and prepends the `i/n` coordinate. This first version includes **no randomization**: the distribution is fully deterministic, `seed` has no effect on the generated points, and `dimension` must be a scalar int (unlike `Halton`, the `i/n` coordinate has no associated prime index). Exported via `qmcpy/discrete_distribution/__init__.py` alongside the other discrete distributions.
- Commands run:
  - `python -m pytest test/test_discrete_distribs.py -v`
  - `python -m pytest test/ -q` (full suite)
  - `python -m doctest qmcpy/discrete_distribution/hammersley.py -v`
  - `mkdocs build --strict`
  - `make doc`

## AI Assistance
- [ ] No substantive AI assistance was used for this PR.
- [x] AI assistance substantively affected this PR, and I describe that use below.

AI tools and affected areas: I used an AI assistant (Claude) throughout this PR, but I directed the design decisions and verified every change myself (running tests, doctests, and reproducing behavior independently rather than accepting output as-is), including catching several issues in my own drafts along the way.

- [x] Independent verification performed.

I ran the full test suite and doctests myself after each change and confirmed all pass locally.

## Checklist
- [x] I linked the relevant issue or explained why none was needed.
- [x] I added or updated tests, docs, notebooks, or explained why they were not needed.
- [x] I verified any equations, citations, benchmarks, numerical claims, or external references changed in this PR.
- [x] I reviewed AI-assisted content for licensing, provenance, attribution, confidentiality, and security concerns.
- [x] I described any CI, dependency, notebook runtime, or generated artifact changes if applicable.

No new dependencies are introduced by this PR. No new data files are added . 

Closes #572

### CODE REVIEW

**Comment by fjhickernell** [comment]
Hi Samuel, welcome to our team.  You have been very active recently.

We are glad to have collaborators from all over.  Would you please tell us a bit about yourself?  

Forgive the long message, but I see that you have your own QMC Python library.  Are you looking to port some of your code to ours?

As Aleksei might have mentioned, we favor extensible sequences, but fixed sample size does make sense if there is a good use case.  Would you mind giving us some motivation?

Also would you motivate the other pull requests?  Korobov points are an interesting idea but the space of possible generators is somewhat limited.  Do you see an advantage in some applications?

If it would help, perhaps a few of us could get on a call to understand what you are trying to do.

If I understand the motivation better it would help me give a better review.

Best,
Fred

**Comment by Samuel-Vangu** [comment]
Hi Fred,

Thanks for the warm welcome and for taking the time to look into this.

I'm not trying to port code over. I built that library during a
research internship on numerical integration, which is where I got
hands-on experience with QMC methods. I wanted to put that experience to
use by contributing to an existing open source project, and QMCPy felt
like a natural fit.

KorobovLattice: you're right that the generator space is limited compared
to CBC. My motivation isn't that Korobov beats CBC — it's more that
Korobov-type lattices are still actively studied today (there are recent papers , e.g Dick & Pillichshammer, March 2026). 

RandomizedLHS: I think it's a good addition on its own merits. Beyond the
basic stratification benefit, there's a classical result (Stein, 1987)
showing it provably beats plain Monte Carlo for functions that are mostly
"additive" in their variables — not just an empirical trend.

HammersleyPointSet: yes, it's a fixed-size construction, but it has a
provably better error bound than plain Halton for that reason. Beyond
that, I just think it's good to offer a wider range of sequences overall.

**Comment by Samuel-Vangu** [comment]
@fjhickernell ,

happy to hop on a call, just let me know a day and time that works for you.

**Comment by fjhickernell** [comment]
How about tomorrow, Thursday, July 30, Noon Chicago / 7 PM Central Europe?  I sent a calendar invite.

**Comment by Samuel-Vangu** [comment]
@fjhickernell
Sounds good, see you tomorrow!

**Comment by fjhickernell** [comment]
So sorry.  Aleksei has a conflict.  How about Monday, August 3, 10 AM Chicago, 5 PM Europe?

**Comment by Samuel-Vangu** [comment]
@fjhickernell
No problem, works for me — see you Monday!

**Comment by Samuel-Vangu** [comment]
Hi @fjhickernell  @alegresor ,I pushed some of thechanges we talked about for Hammersley:

- Renamed the class from `HammersleyPointSet` to `Hammersley`.
- Removed the `alpha` and `base_generating_matrices` parameters (the defaults
  already used internally by `Halton` are sufficient here).
- Updated everything else to match (docstrings, tests, etc.).

Let me know if I'm missing anything else from what we discussed, or if you'd like
me to adjust anything further.

File: qmcpy/discrete_distribution/digital_net_any_bases/hammersley.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        Annals of the New York Academy of Sciences, 86(3):844-874. 1960.
+    """
+
+    def __init__(self,
+                 dimension=1,
+                 seed=None,
+                 t=None,
+                 n_lim=2**32,
+                 warn=True):
+
```
Comment (alegresor) [inline qmcpy/discrete_distribution/digital_net_any_bases/hammersley.py:96]: add docstrings, see other discrete distrib object

**Review by alegresor — CHANGES_REQUESTED** [review]

**Comment by alegresor** [comment]
Closing for #582
